src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -588,6 +590,10 @@ impl TargetInfo {
             .iter()
             .any(|sup| sup.as_str() == split.as_str())
     }
+
+    pub fn get_target_envs(&self) -> &BTreeMap<String, Option<OsString>> {
+        return self.crate_type_process.get_envs();
```

Comment:
Accessing `crate_type_process` seems like a hacky shortcut to me. The process is for learning target info from rustc, but fingerprint is for actual rustc compilation. One extension point for putting it here is for supporting <https://github.com/rust-lang/cargo/issues/10273>, though I believe that will need other rounds of env var system overhual in Cargo.

Reply:
I would agree, the chance of this having unintended consequences seems kinda high since it isn't obvious when looking at `TargetInfo` that this would also influence build-script environment fingerprinting.

Would it be possible to rewrite `apply_env_config` so that there is a separate function that yields a map of resolved config values? Then the code in the fingerprint module could call that and collect the results into a map that it could then pass into the closure.